core/src/main/java/com/xmartlabs/bigbang/core/controller/SharedPreferencesController.kt

```diff
@@ -0,0 +1,104 @@
+package com.xmartlabs.bigbang.core.controller
+
+import android.content.SharedPreferences
+import android.support.annotation.CheckResult
+import com.google.gson.Gson
+import javax.inject.Inject
+import kotlin.reflect.KClass
+
+/**
+ * Manage the [SharedPreferences] storage of the application.
+ *
+ * The entities will be stored in the [SharedPreferences] as a String.
+ * Thus, the entities objects are serialized using [Gson].
+ * The entities are retrieved once from disk and then kept in memory for faster access.
+ */
+open class SharedPreferencesController @Inject
+constructor(private val gson: Gson, private val sharedPreferences: SharedPreferences) : Controller() {
+  private val cachedEntities = HashMap<String, Any>()
+
+  /**
+   * Retrieves the current stored [T] entity, if it exists.
+   *
+   * Only upon first request the [T] entity object will be queried from [SharedPreferences].
+   * Then, it will be stored in memory for faster access.
+   *
+   * @param key  the key of the entity.
+   * *
+   * @param type the type of the entity.
+   * *
+   * @return the current [T] entity if exists.
+   */
+  @CheckResult
+  open fun <T : Any> getEntity(key: String, type: KClass<out T>) =
+    getEntityFromCachedElements(key, type) ?: getEntityFromSharedPreferences(key, type)
+
+  /**
+   * Retrieves the current stored [T] entity, from the cache.
+   *
+   * @param key  the key of the entity.
+   * *
+   * @param type the type of the entity.
+   * *
+   * @return the current [T] entity if exists.
+   */
+  @CheckResult
+  private fun <T : Any> getEntityFromCachedElements(key: String, type: KClass<out T>): T? {
+    val element = cachedEntities[key]
+    return element?.let {
+      @Suppress("UNCHECKED_CAST")
+      if (it.javaClass == type::class) this as? T else null
```

Comment:
No because `it` is an Any object and it cannot be type checked agains a variable parameter (which is `type`)